docker/serverless/gcloud_build_image

```diff
@@ -12,31 +12,65 @@
 #   gcloud endpoints services list
 #   gcloud endpoints configs list --service=${SERVICE}
 # Use the latest one for the CONFIG_ID
+#
+# The script will use the latest ESPv2 version by default.
+# Use the -v option to pass in a custom version (example: -v 2.9).
 
-# Default to the release image if not specified
-BASE_IMAGE=gcr.io/endpoints-release/endpoints-runtime-serverless:2
+# Default to the latest released ESPv2 version.
+BASE_IMAGE_NAME="gcr.io/endpoints-release/endpoints-runtime-serverless"
+ESP_TAG="2"
 
 function error_exit() {
   # ${BASH_SOURCE[1]} is the file name of the caller.
   echo "${BASH_SOURCE[1]}: line ${BASH_LINENO[0]}: ${1:-Unknown Error.} (exit ${2:-1})" 1>&2
   exit ${2:-1}
 }
 
-while getopts :c:s:p:i: arg; do
+while getopts :c:s:p:v:i: arg; do
   case ${arg} in
     c) CONFIG_ID="${OPTARG}";;
     s) SERVICE="${OPTARG}";;
     p) PROJECT="${OPTARG}";;
-    i) BASE_IMAGE="${OPTARG}";;
+    v) ESP_TAG="${OPTARG}";;
+    i)
+      BASE_IMAGE="${OPTARG}"
+      ESP_FULL_VERSION="custom"
+      ;;
     \?) error_exit "Unrecognized argument -${OPTARG}";;
   esac
 done
 
 [[ -n "${PROJECT}" ]] || error_exit "Missing required PROJECT"
 [[ -n "${SERVICE}" ]] || error_exit "Missing required SERVICE"
 [[ -n "${CONFIG_ID}" ]] || error_exit "Missing required CONFIG_ID"
+
+# If user did not pass in custom image, then form the fully-qualified base image.
+if [ -z "${BASE_IMAGE}" ]; then
+  BASE_IMAGE="${BASE_IMAGE_NAME}:${ESP_TAG}"
+fi;
 echo "Using base image: ${BASE_IMAGE}"
 
+# If user did not pass in a custom image, then determine the ESP version.
+if [ -z "${ESP_FULL_VERSION}" ]; then
```

Comment:
Now the logic in this script is complicated. I wonder if we need some automated tests for this. This logic is never tested in our e2e tests, because we always deploy with the `-i` flag.

@qiwzhang @TAOXUY what do you think? Not sure what the best way to test this is.

Reply:
Prefer to add e2e for this piece of  code path since they are heavily used for our users.

Reply:
It is tricky.  Our e2e tests not use official released tag.  only use dev images with SHA.  Not easy to test it as end users 
can we re-org the script into some functions, and use a separate script to test these functions.  only test generated image name, not e2e test?

Reply:
Discussed offline with Wayne. Difficult to unit test because we still rely on GCR to provide tags.

I'll add a quick e2e test for this instead, should be fairly easy. And it will emulate how a user will use the script.

Reply:
SG

Reply:
Updated with test.